### PR TITLE
Manage edit documentation 

### PR DESCRIPTION
**Pull request by lyndaidaii**
Summary of the changes:

Documentation UI change  in manage package page:

* Enable to edit legacy readme file 
* Disable edit embedded readme file 

NuGet Gallery _ Manage My Package

**Question for discussion:** 

Feature flag looks like could be removed here.
For case like we turn off embedded feature, there are legacy and embedded readme, we still don't allow user to edit embedded readme. 

Addresses https://github.com/NuGet/Engineering/issues/3307

### CODE REVIEW

File: src/NuGetGallery/Views/Packages/Manage.cshtml [hunk 1]
```diff
@@ -52,12 +52,15 @@
...
                 expanded: false)
 
-            @ViewHelpers.Section(
-                this,
-                "Documentation",
-                "Documentation",
-                @<text>@Html.Partial("_ManageDocumentation", Model)</text>,
-                expanded: false)
+@*            @if (!Model.IsReadmeFileUploaded)
+            { *@
```
Comment (agr) [inline src/NuGetGallery/Views/Packages/Manage.cshtml:56]: Is that supposed to be commented?

File: src/NuGetGallery/Views/Packages/_EditForm.cshtml [hunk 2]
```diff
@@ -3,7 +3,19 @@
 
     <div id="import-readme-container">
     </div>
-    @Html.Partial("_ImportReadMe")
+@*    @if (Model.VersionSelectList.Selected)
+    {*@
```
Comment (agr) [inline src/NuGetGallery/Views/Packages/_EditForm.cshtml:7]: Same here, remove?

File: src/NuGetGallery/Views/Packages/Manage.cshtml [hunk 3]
```diff
@@ -58,6 +58,7 @@
                 "Documentation",
                 @<text>@Html.Partial("_ManageDocumentation", Model)</text>,
                 expanded: false)
+
```
Comment (lyndaidaii) [inline src/NuGetGallery/Views/Packages/Manage.cshtml:61]: remove this extra line

File: src/NuGetGallery/Views/Packages/_EditForm.cshtml [hunk 4]
```diff
@@ -1,4 +1,5 @@
-﻿<form aria-label="Edit your package" id="edit-readme-form">
+﻿
```
Comment (lyndaidaii) [inline src/NuGetGallery/Views/Packages/_EditForm.cshtml:1]: remove extra line here

File: src/NuGetGallery/ViewModels/TrivialPackageVersionModel.cs [hunk 5]
```diff
@@ -20,5 +21,6 @@ public TrivialPackageVersionModel(string id, string version)
 
         public string Id { get; set; }
         public string Version { get; set; }
+        public bool HasEmbeddedReadme { get; set; }
```
Comment (lyndaidaii) [inline src/NuGetGallery/ViewModels/TrivialPackageVersionModel.cs:24]: This is new attribute introduced here

**Comment by lyndaidaii** [comment]
Please ignore previous comments. This is new different UI behavior comparing previous one

File: src/NuGetGallery/Scripts/gallery/page-edit-readme.js [hunk 6]
```diff
@@ -184,6 +185,23 @@
...
+
+            if (model.IsEmbeddedReadmeEnabled) {
+                var selectedVersion = _selectVersion.val();
+
+                if (!selectedVersion) {
+                    // No version is selected.
+                    return;
+                }
+
+                if (model.Versions[selectedVersion].HasEmbeddedReadme) {
```
Comment (lyndaidaii) [inline src/NuGetGallery/Scripts/gallery/page-edit-readme.js:197]: It hidden edit/submit button when there is embeddedReadme and embeddedReadme feature flag is on. I am considering to remove the feature flag, looks like it's not necessary when roll back case.

File: src/NuGet.Services.Entities/Package.cs [hunk 7]
```diff
@@ -156,6 +156,18 @@ public bool HasReadMe
             }
         }
 
+        /// <summary>
+        /// Signifies whether or not the embedded ReadMe exists
+        /// </summary>
+        [NotMapped]
+        public bool HasEmbeddedReadMe
```
Comment (loic-sharma) [inline src/NuGet.Services.Entities/Package.cs:163]: I know our codebase is inconsistent between `ReadMe` and `Readme`, but from my understanding we would like to migrate to the `Readme` casing. If so, should we use the `Readme` casing on all new properties/methods you introduce?

```suggestion
        public bool HasEmbeddedReadme
```

File: src/NuGetGallery/Helpers/ViewModelExtensions/ManagePackageViewModelFactory.cs [hunk 8]
```diff
@@ -157,5 +152,21 @@ public ManagePackageViewModelFactory(IIconUrlProvider iconUrlProvider)
...
+            TrivialPackageVersionModel model,
+            UrlHelper url)
+        {
+            var submitUrlTemplate = url.PackageVersionActionTemplate("Edit");
+            var getReadMeUrlTemplate = url.PackageVersionActionTemplate("GetReadMeMd");
+
+            var result = new ManagePackageViewModel.VersionReadMeState(
+                submitUrlTemplate.Resolve(model),
+                getReadMeUrlTemplate.Resolve(model),
+                null);
```
Comment (skofman1) [inline src/NuGetGallery/Helpers/ViewModelExtensions/ManagePackageViewModelFactory.cs:166]: >null [](start = 16, length = 4)

use named params for readability

**Review by skofman1 — APPROVED** [review]
:shipit:

File: src/NuGet.Services.Entities/Package.cs [hunk 9]
```diff
@@ -156,6 +156,18 @@ public bool HasReadMe
             }
         }
 
+        /// <summary>
+        /// Signifies whether or not the embedded ReadMe exists
```
Comment (lyndaidaii) [inline src/NuGet.Services.Entities/Package.cs:160]:
```suggestion
        /// Signifies whether or not the embedded Readme exists
```

File: src/NuGetGallery/Scripts/gallery/page-edit-readme.js [hunk 10]
```diff
@@ -184,6 +185,22 @@
             if (model === null || !model.IsSymbolsPackage) {
                 BindReadMeDataManager.bindReadMeData(model);
             }
+
+
```
Comment (lyndaidaii) [inline src/NuGetGallery/Scripts/gallery/page-edit-readme.js:189]: remove extra line here

File: src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs [hunk 11]
```diff
@@ -38,6 +38,7 @@ public class DisplayPackageViewModel : ListPackageItemViewModel
         public bool IsPackageDependentsEnabled { get; set; }
         public NuGetPackageGitHubInformation GitHubDependenciesInformation { get; set; }
         public bool HasEmbeddedIcon { get; set; }
+        public bool HasEmbeddedReadmeFile { get; set; }
```
Comment (zhhyu) [inline src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:41]: Do we use this field anywhere?

Comment (lyndaidaii) [reply src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:41]: Good catch, it needs to be deleted

File: src/NuGetGallery/Views/Packages/_ManageDocumentation.cshtml [hunk 12]
```diff
@@ -62,7 +62,7 @@
             </div>
         </div>
 
-        @Html.Partial("_EditForm")
+        @Html.Partial("_EditForm", Model)
     </div>
```
Comment (zhhyu) [inline src/NuGetGallery/Views/Packages/_ManageDocumentation.cshtml:66]: Do we really need to pass the model here?

File: src/NuGetGallery/ViewModels/ManagePackageViewModel.cs [hunk 13]
```diff
@@ -50,6 +50,7 @@ public VersionReadMeState(string submitUrl, string getReadMeUrl, string readMe)
             public string SubmitUrl { get; }
             public string GetReadMeUrl { get; }
             public string ReadMe { get; }
+            public bool HasEmbeddedReadme { get; set; }
```
Comment (zhhyu) [inline src/NuGetGallery/ViewModels/ManagePackageViewModel.cs:53]: Do you think whether it's better if we also set this value in the constructor, and restrict it to a read-only field? I see that we only use this field in the corresponding JS file, so it should be safe to do so?

**Comment by zhhyu** [comment]
Let's verify the behavior manually, as we don't have a good test coverage on views. :shipit:

File: src/NuGetGallery/Scripts/gallery/page-edit-readme.js (hunk 6 above)
Comment (zhhyu) [reply src/NuGetGallery/Scripts/gallery/page-edit-readme.js]: I agree. There is no need to use the feature flag here. The feature flag is used to block customers to upload a package with an embedded readme file. Here checking the DB flag is enough.

**Pull request merged**